shmux.1
.TH SHMUX 8 "2017-01-13"
.DA January 13, 2017
.SH NAME
shmux - Shell Multiplexor
.SH SYNOPSIS
.B shmux
[
.B -h
|
.B -V
]

.B shmux
[
.B -bBdFmpqQstv
] [
.B -C \fItimeout\fP
] [
.B -M \fImax\fP
] [
.B -r \fIrcmd\fP
] [
.B -S \fImode\fP
] [
.B -e \fIlist\fP
] [
.B -E \fIlist\fP
] [
.B -a \fIanalyzer\fP
] [
.B -A \fIcondition\fP
] [
.B -o \fIdir\fP
] [
.B -P \fItimeout\fP
] [
.B -T \fItimeout\fP
]
.B -c \fIcommand\fP
[ - | \fItargets...\fP ]

.SH DESCRIPTION
\fBshmux\fP is program for executing the same \fIcommand\fP on many hosts
in parallel.  For each target, a child process is spawned by \fBshmux\fP,
and a shell on the target obtained one of the supported methods: \fIrsh\fP,
\fIssh\fP, or \fIsh\fP.  The output produced by the children is received by
\fBshmux\fP and either (optionally) output in turn to the user, and/or
written to files for later processing.  The list of \fItargets\fP may
either be specified on the command line or be read from the standard input.

The default method may be set by using the \fB-r\fP option, and may be
overridden for any target by prefixing the target name by the method and a
colon.  Two special methods, \fIssh1\fP and \fIssh2\fP may be used to
select the SSH protocol version 1 or 2, respectively.  For the \fIsh\fP
method, it is up to the specified \fIcommand\fP to contact the target.  In
order for the executed \fIcommand\fP to know which target it is being
invoked for, the \fISHMUX_TARGET\fP environment variable is set by
\fBshmux\fP.

Before executing the specified \fIcommand\fP, \fBshmux\fP will optionally
ping each target to ensure that it can be reached, and/or run a dummy test
\fIcommand\fP to make sure that the target not only is alive, but that it
is possible to cleanly execute a command on it.  Both these tests are
typically run with a fairly short timeout to quickly dismiss unavailable
targets rather than waiting for the standard (longer) network timeout.

After executing the specified \fIcommand\fP, \fBshmux\fP will optionally
process the output or run a user specified command to determine whether the
\fIcommand\fP was successful or resulted in an error.  Optionally, the
\fIcommand\fP output is suppressed for successful targets.

Finally \fBshmux\fP will print out a summary of the results and list
targets with errors.

.SH OPTIONS
.IP "\fB-h\fP"
Display a brief help message.
.IP "\fB-V\fP"
Display the version information.
.IP "\fB-c \fIcommand\fP"
Specify the \fIcommand\fP to execute on targets.
.IP "\fB-C \fItimeout\fP"
Specify a timeout for the command being executed on targets.  This should
be a number followed by a time unit.  The following are valid time units:
s(econds), m(inutes), h(our), d(ays), w(eeks).  See the \fIPROCESS
MANAGEMENT\fP section for details on how this is handled.
.IP "\fB-M \fImax\fP"
Defines the maximum number of spawned processes.  While there is no real
(or hard coded) limitation for this, the system resources are typically
bounded and this affects \fBshmux\fP in turn.  The most critical resource
is the maximum number of open files a process may have.  But one should
also consider the load imposed on the system by the creation of a
(potentially) large number of processes.
.IP "\fB-r \fIrcmd\fP"
Defines the default method used to run a shell on targets.
.IP "\fB-S \fImode\fP"
This options defines the strategy used by \fBshmux\fP when spawning
children to run the \fIcommand\fP on targets.  If the \fImode\fP is set to
"all", then processes will be spawned as fast as possible.  When set to
"check", processes will be spawned as long as \fIcommand\fPs succeed.
Finally, the default \fImode\fP, "one", is to spawn only one process at
first, and wait for it to succeed, at which point \fBshmux\fP will switch
to the "check" mode.  In the "one" or "check" modes, when the \fIcommand\fP
fails on a target, \fBshmux\fP will stop spawning new processes.
.IP "\fB-F\fP"
When using either the "one" or "check" spawn modes (see \fB-S\fP),
\fBshmux\fP will stop spawning new processes if the \fIcommand\fP fails on
a target.  By default, \fBshmux\fP then waits for user input if available,
but reverts to spawning processes if such input cannot be obtained (because
the interactive mode is disabled, either via \fB-B\fP or lack of a
terminal).  Using the \fB-F\fP option changes this behaviour to force
\fBshmux\fP to quit once all existing processes have terminated after such
a failure.
.IP "\fB-e \fIlist\fP"
Defines which \fIcommand\fP exit codes should be considered errors and
reported as such.  The \fIlist\fP should be a comma separated list of
ranges of the form "x", "x-y", "x-" or "-y" where "x" and "y" are numbers
between 0 and 255.
.IP "\fB-E \fIlist\fP"
Defines which \fIcommand\fP exit codes should always be output to the
screen.  Codes that are considered errors (see the \fB-e\fP option) are
always shown, but by default, other exit codes are only shown if the
\fB-v\fP option is specified.  This allows more flexibility.
.IP "\fB-a \fIanalyzer\fP"
Defines how output should be analyzed by \fBshmux\fP after the
\fIcommand\fP completes on a target.  By default, nothing is done.  Valid
options are: \fIlnregex\fP, \fIlnpcre\fP, \fIregex\fP, \fIpcre\fP and
\fIrun\fP.  This option requires \fB-o\fP to be used as well.
.IP "\fB-A \fIcondition\fP"
When the \fB-a\fP option is used, it is also necessary to configure the
chosen analyzer.

For line based regular expressions (\fIlnregex\fP and \fIlnpcre\fP), the
\fB-A\fP option should be specified twice:  once for the standard output
and once for the standard error output.  In both cases, the argument should
be the name of a file containing a list of regular expressions, one per
line.  Each regular expression must be prefixed by the character '=' or '!'.
The output of the \fIcommand\fP is analyzed line per line by checking
whether it matches any of the specified regular expressions.  If the first
matching regular expression is prefixed by '=', then the \fIcommand\fP is
considered to be successful (so far).  Any other result (e.g. no match, or
a match prefixed by '!') indicates an error condition.  The default is to
consider standard output normal, and standard error output indicative of an
error.

For other regular expressions (\fIregex\fP and \fIpcre\fP), the \fB-A\fP
option should be specified twice: once for the standard output and once for
the standard error output.  This option allows defining the expression to
be applied to the output.  If the \fIcondition\fP starts with the character '='
then a match indicates that the \fIcommand\fP completed without error.
If the \fIcondition\fP starts with the character '!' then a failure to
match indicates that the \fIcommand\fP completed without error.  The second
character of the condition should either be '=' if the regular expression
follows or '<' if the regular expression should be read from the file which
name follows.  (The '=' characters may be omitted.)  A \fIcondition\fP must
be specified for the standard output, but is optional for standard error
output.  The default for the latter is to consider any output as indicative
of an error.

For the \fIrun\fP analyzer, the \fB-A\fP must be specified at least once
with the name of a program to run, and optionally a second time to specify
a timeout which defaults to 15 seconds if unset.  This timeout is
implemented using alarm(3) and will not otherwise be enforced by
\fBshmux\fP.  The program will be called after a \fIcommand\fP completes on
a target unless the \fIcommand\fP exit code is considered an error (see
\fB-e\fP) with two arguments: the target name and the directory specified
with the \fB-o\fP option.  The output of the program will always be shown
to the user.  Exit codes other than 0 indicate that the \fIcommand\fP
failed for the particular target.
.IP "\fB-o \fIdir\fP"
If specified, \fBshmux\fP will place the output and (for normal
terminations) the exit code of the executed commands in files under this
directory.  For each target, the files will be named after the target name
so as to make it easy to identify.  The files must not already exist, so it
is recommended that the directory be empty.  This also means that each
target must be unique.  The directory will be created if it does not
already exist.
.IP "\fB-p\fP"
Ping targets to verify they are alive before doing anything.  The target
names must be unique or bad things will happen.
.IP "\fB-P \fItimeout\fP"
Defines the initial target ping timeout in milliseconds, see
\fIfping(8)\fP.  (Implies \fB-p\fP.)
.IP "\fB-t\fP"
Before executing the specified \fIcommand\fP for a target, send a simple
"echo" test and verify that the output is correct.  This step goes beyond
the simple ping test as it verifies that it is possible to get a shell on
the target.  If this option is specified twice (or once when \fB-T\fP is
used), output for failed tests will be displayed to help the user
understand what went wrong.
.IP "\fB-T \fItimeout\fP"
Defines the test timeout in seconds.  (Implies \fB-t\fP.)
.IP "\fB-m\fP"
By default the output is displayed as soon as it is received.  For
multi-line outputs, this will typically result in output from several
targets being mixed.  This option may be used to keep each target output
separate.
.IP "\fB-b\fP"
Show the bare output from the executed commands instead of prefixing each
line by the corresponding target name.
.IP "\fB-B\fP"
Enables batch mode which forcefully disables the interactive mode, ignoring
input from the terminal.  See also \fB-s\fP and \fB-F\fP.
.IP "\fB-s\fP"
Suppress the progress status line.  See also \fB-B\fP.
.IP "\fB-q\fP"
Used once, this will suppress the output from successful targets (as
defined by the use of options options \fB-e\fP, \fB-a\fP and \fB-A\fP.  If
using the \fIrun\fP analyzer, the output of all targets is suppressed.
Note that using this option effectively implies \fB-m\fP in most cases as
failure is ultimately determined upon completion of the \fIcommand\fP.
(Only the \fIlnregex\fP and \fIlnpcre\fP analyzers are able to determine
errors while the \fIcommand\fP is running.)  When specified twice, this
option allows to completely suppress output from the \fIcommand\fP run on
targets.  This also changes the default spawn stategy to "all".  Using this
option (once or twice) requires using \fB-o\fP.
.IP "\fB-Q\fP"
Suppress the final summary of results.
.IP "\fB-v\fP"
Display internal status messages.
.IP "\fB-D\fP"
Display internal debug messages.

.SH EXIT CODES
\fBshmux\fP will optionally report the exit code of \fIthe command it
runs\fP for each target.  For the \fIsh\fP command, this will be the same
as the \fIcommand\fP itself, but this is the exception.  For other methods,
the exit code reported by \fBshmux\fP really is the code returned by the
program used to spawn \fIcommand\fP for that method.

In particular, \fIrsh\fP typically only returns either 0 or 1 while
different implementations of \fIssh\fP have different behaviours which may
include returning the actual exit code of the remotely executed command or
not.

.SH PROCESS MANAGEMENT
Children processes spawned by \fBshmux\fP are made process group leaders
(using setpgid(2)) in order to group all of their descendants in a single
process group which allows to easily send signals to all of these
processes.  This is especially useful for the \fIsh\fP method which can
generate many processes locally, but has no effect on remote processes.

\fBshmux\fP uses pipe(2) to capture the output generated by its children
(even when this output is redirected to a file with \fB-o\fP) and will read
from these until they are closed (by the child as well as its descendants,
if any).  Once there is nothing left to read, \fBshmux\fP will wait for the
child and its descendants to terminate before considering that particular
target done with.

The timeout is implemented by using alarm(3) which (unless canceled or
intercepted) will terminate the child process upon receipt but is not
inherited by descendants of that process.  For that reason, \fBshmux\fP
will send a SIGALRM signal to the child's process group when the child
itself terminates upon such signal.

After the scheduled SIGALRM time, \fBshmux\fP will wait an additional 5
seconds for all of the processes in the group to terminate to give the
child a chance to handle the timeout more gracefully if desired.  It will
then send a SIGTERM signal to the child's process group.  After another 5
seconds, if any process is still alive in the group, a final SIGKILL signal
is sent to the process group.

Note that only commands which terminate because of an ALRM signal or
because of a TERM or KILL signal after such signal has been issued by
\fBshmux\fP will be reported as having timed out in the final summary
report produced by \fBshmux\fP upon exiting.  This is not as accurate as it
should be.

.SH INTERACTIVE MODE
By default, \fBshmux\fP offers a minimal "interactive mode" while running:
it reads commands from the terminal and acts upon them accordingly.  This
can be disabled using the \fB-B\fP option.  The following commands are
currently recognized:

.IP "\fBq\fP"
Prevent \fBshmux\fP from spawning any more children, wen for active
children to complete, and then quit.
.IP "\fBQ\fP"
Quit immediately (e.g. abort), not waiting for any active children to
terminate.  This isn't recommended as children may be left running
unattended.  A better alternative is to press control-C which will be
received by \fBshmux\fP and all running children.  These will typically
abort, and \fBshmux\fP terminate in a cleaner fashion.
.IP "space"
Initiate a pause, this prevents \fBshmux\fP from spawning any more children
to do its work.  The following three commands may be used to resume work.
.IP "1"
This is equivalent to specifying "\fB-S\fP \fImore\fP" on the command line,
but may be used at any time.
.IP "return"
This is equivalent to specifying "\fB-S\fP \fIcheck\fP" on the command line,
but may be used at any time.
.IP "+"
This is equivalent to specifying "\fB-S\fP \fIall\fP" on the command line,
but may be used at any time.
.IP "F"
This toggles the failure mode for the "one" and "check" spawn modes (see
\fB-S\fP) between "pause" and "quit" (see \fB-F\fP).
.IP "\fBp\fP"
Display the list of pending targets.
.IP "\fBr\fP"
Display the list of targets for which a children process is currently
running.
.IP "\fBf\fP"
Display the list of targets for which \fBshmux\fP failed to run the
\fIcommand\fP.
.IP "\fBe\fP"
Display the list of targets for which the \fIcommand\fP completed with
error(s).
.IP "\fBs\fP"
Display the list of targets for which the \fIcommand\fP completed
successfully.
.IP "\fBa\fP"
Display the list of all targets with their status.
.IP "\fBk\fP"
Allows sending a signal to a target.
.IP "\fBv\fP"
Toggles whether internal messages are displayed or not.
.IP "\fBD\fP"
Toggles whether debug messages are displayed or not.

.SH ENVIRONMENT VARIABLES
\fBshmux\fP will use the following environment variables if set:

.IP SHMUX_MAX
Specifies the maximum number of processes to spawn simultaneously.  If the
\fB-M\fP option is specified, it overrides this variable.
.IP SHMUX_RCMD
Specifies the default command used to run a shell on targets.  If the
\fB-r\fP option is specified, it overrides this variable.
.IP SHMUX_SH
Specify the shell to use for targets using the \fIsh\fP method.  (Default
is "/bin/sh".)
.IP SHMUX_RSH
Specify an alternative command to "rsh" to use for targets using the
\fIrsh\fP method.
.IP "SHMUX_SSH, SHMUX_SSH1, SHMUX_SSH2"
Specify an alternative command to "ssh" to use for targets using the
\fIssh\fP, \fIssh1\fP and \fIssh2\fP method (respectively).  The variable
\fISHMUX_SSH\fP is used for all methods, but superseded by the other
specific variables.
.IP "SHMUX_SSH_OPTS, SHMUX_SSH1_OPTS, SHMUX_SSH2_OPTS"
The variables may be used to specify replace the default "-xa
-oLogLevel=ERROR" options passed to "ssh", "ssh1" and "ssh2"
(respectively).  The variable \fISHMUX_SSH_OPTS\fP is used for all
methods, but superseded by the other
specific variables.

Multiple arguments may be supplied, using white-space characters to
separate them from each other.  You can use single- and/or double-quotes to
preserve white-spaces in an argument (but not backslashes).  Note that
doubling  a single-quote inside a single-quoted string gives you a
single-quote; likewise for double-quotes (though you  need  to  pay
attention  to which  quotes  your  shell  is parsing and which quotes
\fBshmux\fP is parsing).
.IP SHMUX_SPAWNMODE
Specifies the default spawn mode.  If the \fB-S\fP option is specified, it
overrides this variable.
.IP SHMUX_ERRORCODES
Specifies the default list of exit codes to be considered errors.  If the
\fB-e\fP option is specified, it overrides this variable.
.IP SHMUX_SHOWCODES
Specifies the default list of exit codes to be always display.  If the
\fB-E\fP option is specified, it overrides this variable.

.SH EXIT STATUS
The following exit values are returned:
.IP 0
indicates that \fBshmux\fP was able to complete its work.  However, this
does not mean that all targets successfully ran the \fIcommand\fP, nor that
\fBshmux\fP ran without any error.
.IP 1
indicates \fBshmux\fP was cleanly interrupted before it could finish its
work.
.IP 2
indicates \fBshmux\fP aborted upon user request, possibly leaving some
children running.
.IP 3
indicates \fBshmux\fP encountered a fatal error which caused it to abort,
possibly leaving some children running.
.IP 4
indicates a syntax error or a fatal error which caused \fBshmux\fP to abort
before it could spawn any children.

.SH SEE ALSO
.IR fping (8),
.IR sh (1),
.IR rsh (1),
.IR ssh (1),
.IR regex (3),
.IR pcre (3).

.SH AVAILABILITY
The latest official release of \fBshmux\fP is available on GitHub at
http://github.com/shmux.

.SH AUTHOR
Christophe Kalt <ck@taranis.org>

.SH BUGS

CHANGES
Changes since 1.0.2 [2008-12-21]:

- Default to using -oLogLevel=ERROR with ssh

Changes since 1.0.1 [2006-08-30]:

- Mac OS X port (with help from Steve Huff).
- use safe string functions (courtesy of Olivier Cherrier).

Changes since 1.0 [2006-08-30]:

- don't core dump if given host list is empty (reported by Brian King).
- fixed host name length measurement when reading from stdin.

No change since 1.0b10 [2006-08-10].

Changes since 1.0b9 [2006-07-06]:

- new 'F', 'v' and 'D' commands for interactive mode.
- added -F option (as suggested by Chase Phillips).
- fixed long standing but rare data corruption bug (reported by Jon Amundsen).

Changes since 1.0b8 [2006-06-30]:

- child exit status was not saved when waiting for its orphans.
- look for termcap in ncurses library (reported by Stephan Hegel).

Changes since 1.0b7 [2006-05-22]:

- SHMUX_SSH*_OPTS are now parsed by shmux, allowing passing more complex
  options strings.  (loosely based on code from Jos Backus.)
- added -B option (courtesy of Chase Phillips and Jos Backus).
- minor enhancement of "interactive mode" status reporting.

Changes since 1.0b6 [2004-12-15]:

- improved timeout enforcement using process groups (suggested by Jos Backus).
- improved handling of "impossible" waitpid() failures.

Changes since 1.0b5 [2004-09-20]:

- user@target now works with rsh.
- fixed miscellaneous assumptions/bugs related to fping.
- user@target is no longer incompatible with -p.

Changes since 1.0b4 [2004-08-07]:

- use SHMUX_SPAWNMODE environmental variable, as per the manual page.
- new SHMUX_MAX environmental variable (suggested by Jon Amundsen).

Changes since 1.0b3 [2004-07-06]:

- ported to cygwin (problems reported by Jon Amundsen).
- better background detection code.
 
Changes since 0.13b [2004-06-07]:

- added code to try to avoid hanging accessing the controlling terminal.
- Sun's OpenSSH chokes on "-xa", use "-x -a" instead (reported by Poehner Uwe).

Changes since 0.12b [2004-04-05]:

- clear environment of harmful variables to run tests.
- updated "timers" test for Solaris.
- minor Makefile changes.

Changes since 0.11a [2003-11-28]:

- added test suite.
- fixed various minor problems with parsing of -A for regex/pcre analyzers.
- fixed fatal regex analyzer initialization problem.
- fixed mmap() usage.

Changes since 0.10a [2003-06-22]:

- suppressed extraneous empty lines in final report.
- fixed minor scheduling bug resulting in slower runs with the default "-S one".
- cleaned up handling of early termination.
- minor code cleanup in term.c.
- command timeouts (-C) are now also enforced by shmux with SIGTERM and SIGKILL.

Changes since 0.9a [2003-05-04]:

- plugged termcap escape sequence leakage to output files.
- assertion was no longer valid when using -o but not -qq.
- -qq now implies "-S all" (as default only).
- assertion was failed with -qq.
- updated mcmd.sh for new shmux usage.
- added ln{pc,}re analyzer.

Changes since 0.8a [2003-04-30]:

- fixed -t option behaviour.
- tty_restore() must do its work only for shmux, not forked processes.
- reset signal handlers in exec().
- Added new -q option.
- Renamed -q option to -Q.

Changes since 0.7a [2003-03-30]:

- fixed parsing of ranges so that "x-" actually works.
- added -S option.
- added minimal interactivity.
- cleaned up window width tracking.
- fixed unlikely (so far) bug in target_results().

Changes since 0.6a [2003-03-23]:

- reading the list of targets from the standard input is now supported.
- added missing variable initialization.
- fixed installation of mcmd.sh.
- renamed some functions to avoid names clashes (FreeBSD 4.7).
- simplified term.c system inclusions (RedHat 7.3 has broken includes).

Changes since 0.5a [2003-01-05]:

- added mcmd.sh sample script.
- fixed potential crash in target_results() when interrupted.
- status progress is now truncated to the window width.
- aggressively try to find a tty to display progress status.
- terminal initialization logic is now more flexible.
- added -a and -A options.
- fixed (very rare) file descriptor leak.
- minor fix in algorithm spawning new children should speed things up a bit.
- "exit" files were created for -m rather than -o.
- added a workaround for Linux's pathetic implementation of strtok().

Changes since 0.4a [2002-08-21]:

- fixed the sometimes erroneous "Pending" count in status display.
- -o options now generates more files per target.
- final summary report is now more detailed (in part thanks to -e).
- added -e and -E options.
- IRIX hides basename() in libgen, look there just in case.
- termcap/curses checks and logic cleaned up in configure.ac & term.c.
- MAXNAMLEN replaced with PATH_MAX.

Changes since 0.3a [2002-07-11]:

- improved handling of partial lines.
- truncated lines would cause a crash (failed assertion) with -o.
- added -m option.
- enabled gcc macros for v3.x
- fixed miscellaneous little bugs in exec().
- fixed getopt() usage.

Changes since 0.2a [2002-07-08]:

- SIGINT is now caught, but not by children (which will die right away).
- create the directory supplied with the -o option if it doesn't exist.
- cleaned up environment manipulation in exec().
- -b negates old -v option which is now the default, -v replaces old -d.
- added environment variables SHMUX_SSH*_OPTS.
- added -xa options to ssh by default.
- SHMUX_SH was overloaded, added SHMUX_RCMD, renamed -m option to -r.
- use strsignal(2) instead of sys_signame[].

Changes since 0.1a [2002-07-04]:

- added -o option.
- don't display the status line unless stdout is a tty, even without -s.
- added kludge to work around NetBSD's broken poll(2) implementation.
- adjusted formula deciding the maximum allowed for -M.
- added -C option.
- algorithm now prefers starting real commands over test commands.
- made it easier to understand why tests fail (see -t).
- added environment variables SHMUX_*SH.
- cleaned up loop.c for better readability, also fixed output processing bugs.

src/shmux.c
/*
** Copyright (C) 2002-2008 Christophe Kalt
**
** This file is part of shmux,
** see the LICENSE file for details on your rights.
*/

#include "os.h"

#if defined(HAVE_LIBGEN_H)
# include <libgen.h>
#endif
#include <dirent.h>
#if defined(HAVE_PATHS_H)
# include <paths.h>
#else
# define _PATH_TMP "/tmp"
#endif
#include <time.h>
#include <sys/stat.h>
#if defined(HAVE_PCRE_H)
# include <pcre.h>
#endif

#if !defined(HAVE_BASENAME)
# define basename(s) ((strrchr(s,'/') == NULL) ? (char*)s : (strrchr(s,'/')+1))
#endif

#include "version.h"

#include "analyzer.h"
#include "byteset.h"
#include "loop.h"
#include "target.h"
#include "term.h"
#include "units.h"

static char const rcsid[] = "@(#)$Id$";

extern char *optarg;
extern int optind, opterr;

char *myname;

/* Miscellaneous default settings */
#define DEFAULT_ANALYSIS "regex"
#define DEFAULT_ERRORCODES "1-"
#define	DEFAULT_MAXWORKERS 10
#define DEFAULT_PINGTIMEOUT "500"
#define DEFAULT_RCMD "ssh"
#define DEFAULT_SPAWNMODE "one"
#define DEFAULT_TESTTIMEOUT 15

static void usage(int);

static void
usage(detailed)
int detailed;
{
    fprintf(stderr, "Usage: %s [ options ] -c <command> [ - | <target1> [ <target2> ... ] ]\n", myname);
/*    fprintf(stderr, "Usage: %s [ options ] -i [ <target1> [ <target2> ... ] ]\n", myname);*/
    if (detailed == 0)
	return;
    fprintf(stderr, "  -h            Print this message.\n");
    fprintf(stderr, "  -V            Output version info.\n");
    fprintf(stderr, "\n");
    fprintf(stderr, "  -c <command>  Command to execute on targets.\n");
    fprintf(stderr, "  -C <timeout>  Set a command timeout.\n");
    fprintf(stderr, "\n");
    fprintf(stderr, "  -M <max>      Maximum number of simultaneous processes (Default: %u).\n", DEFAULT_MAXWORKERS);
    fprintf(stderr, "  -r <rcmd>     Set the default method (Default: %s).\n", DEFAULT_RCMD);
    fprintf(stderr, "  -p            Ping targets to check for life.\n");
    fprintf(stderr, "  -P <millisec> Initial target timeout given to fping (Default: %s).\n", DEFAULT_PINGTIMEOUT);
    fprintf(stderr, "  -t            Send test command to verify target health.\n");
    fprintf(stderr, "  -T <seconds>  Time to wait for test answer (Default: %d).\n", DEFAULT_TESTTIMEOUT);
    fprintf(stderr, "\n");
    fprintf(stderr, "  -S <mode>     Spawn strategy (Default: \"%s\")\n", DEFAULT_SPAWNMODE);
    fprintf(stderr, "  -F            Gracefully quit rather than pause\n");
    fprintf(stderr, "  -e <range>    Exit codes to consider errors (Default: \"%s\")\n", DEFAULT_ERRORCODES);
    fprintf(stderr, "  -E <range>    Exit codes to always display (Default: \"%s\")\n", DEFAULT_ERRORCODES);
    fprintf(stderr, "  -a <type>     Analysis type (Default: %s)\n", DEFAULT_ANALYSIS);
    fprintf(stderr, "  -A <test>     Analyze output to determine success from failure.\n");
    fprintf(stderr, "\n");
    fprintf(stderr, "  -o <dir>      Send the output to files under the specified directory.\n");
    fprintf(stderr, "  -m            Don't mix target outputs.\n");
    fprintf(stderr, "  -b            Show bare output without target names.\n");
    fprintf(stderr, "  -B            Batch mode.\n");
    fprintf(stderr, "  -s            Suppress progress status.\n");
    fprintf(stderr, "  -q            Suppress output of successful targets.\n");
    fprintf(stderr, "  -qq           Suppress target output.\n");
    fprintf(stderr, "  -Q            Suppress final summary.\n");
    fprintf(stderr, "  -v            Display internal status messages.\n");
    fprintf(stderr, "  -D            Display internal debug messages.\n");
}

int
main(int argc, char **argv)
{
    int badopt, rc;
    int opt_prefix, opt_status, opt_interactive, opt_quiet, opt_internal, opt_debug;
    int opt_ctimeout, opt_outmode, opt_maxworkers, opt_fail, opt_vtest;
    u_int opt_test, opt_analyzer;
    char *opt_analyze, *opt_outanalysis, *opt_erranalysis;
    char *opt_spawn, *opt_command, *opt_odir, *opt_ping, *opt_rcmd;
    char tdir[PATH_MAX];
    int longest;
    time_t start;

    myname = basename(argv[0]);

    opt_prefix = opt_status = opt_interactive = 1;
    opt_quiet = opt_internal = opt_debug = 0;
    opt_outmode = OUT_MIXED;
    if (getenv("SHMUX_MAX") != NULL)
        opt_maxworkers = atoi(getenv("SHMUX_MAX"));
    else
        opt_maxworkers = DEFAULT_MAXWORKERS;
    opt_ctimeout = opt_fail = opt_test = opt_vtest = 0;
    opt_analyze = opt_outanalysis = opt_erranalysis = NULL;
    opt_command = opt_odir = opt_ping = NULL;
    opt_rcmd = getenv("SHMUX_RCMD");
    opt_spawn = NULL;
    if (getenv("SHMUX_SPAWNMODE") != NULL)
	opt_spawn = getenv("SHMUX_SPAWNMODE");
    if (opt_rcmd == NULL)
	opt_rcmd = DEFAULT_RCMD;
    if (getenv("SHMUX_ERRORCODES") != NULL)
	byteset_init(BSET_ERROR, getenv("SHMUX_ERRORCODES"));
    else
	byteset_init(BSET_ERROR, DEFAULT_ERRORCODES);
    if (getenv("SHMUX_SHOWCODES") != NULL)
	byteset_init(BSET_SHOW, getenv("SHMUX_SHOWCODES"));
    else
	byteset_init(BSET_SHOW, "");

    badopt = 0;
    while (1)
      {
        int c;
	
        c = getopt(argc, argv, "a:A:bBc:C:De:E:FhmM:o:pP:qQr:sS:tT:vV");
	
        /* Detect the end of the options. */
        if (c == -1)
            break;
	
        switch (c)
          {
	  case 'a':
	      opt_analyze = optarg;
	      break;
	  case 'A':
	      if (opt_outanalysis == NULL)
		  opt_outanalysis = optarg;
	      else
		  opt_erranalysis = optarg;
	      break;
	  case 'b':
	      opt_prefix = 0;
	      break;
	  case 'B':
	      opt_interactive = 0;
	      break;
	  case 'c':
	      opt_command = optarg;
	      break;
	  case 'C':
	      opt_ctimeout = unit_time(optarg);
	      break;
	  case 'D':
	      opt_debug = 1;
	      break;
	  case 'e':
	      byteset_init(BSET_ERROR, optarg);
	      break;
	  case 'E':
	      byteset_init(BSET_SHOW, optarg);
	      break;
          case 'F':
              opt_fail = 1;
              break;
          case 'h':
              usage(1);
              exit(RC_OK);
              break;
	  case 'm':
	      opt_outmode &= ~(OUT_NULL|OUT_MIXED);
	      opt_outmode |= OUT_ATEND;
	      break;
	  case 'M':
	      opt_maxworkers = atoi(optarg);
	      break;
	  case 'o':
	      opt_odir = optarg;
	      break;
	  case 'p':
	      if (opt_ping == NULL)
		  opt_ping = DEFAULT_PINGTIMEOUT;
	      break;
	  case 'P':
	      opt_ping = optarg;
	      break;
	  case 'q':
	      if ((opt_outmode & OUT_IFERR) == 0)
		  opt_outmode |= OUT_IFERR;
	      else
		{
		  opt_outmode = OUT_NULL;
		  if (opt_spawn == NULL)
		      opt_spawn = "all";
		}
	      break;
	  case 'Q':
	      opt_quiet = 1;
	      break;
	  case 'r':
	      opt_rcmd = optarg;
	      break;
	  case 's':
	      opt_status = 0;
	      break;
	  case 'S':
	      opt_spawn = optarg;
	      break;
	  case 't':
	      if (opt_test == 0)
		  opt_test = DEFAULT_TESTTIMEOUT;
	      opt_vtest += 1;
	      break;
	  case 'T':
	      opt_test = atoi(optarg);
	      opt_vtest += 1;
	      break;
	  case 'v':
	      opt_internal = 1;
	      break;
	  case 'V':
#if !defined(HAVE_PCRE_H)
	      printf("%s version %s\n", myname, SHMUX_VERSION);
#else
	      printf("%s version %s (PCRE version %s)\n",
		     myname, SHMUX_VERSION, pcre_version());
#endif
	      exit(RC_OK);
	  case '?':
	      badopt += 1;
	      break;
	  default:
	      abort();
	  }
      }

    if (optind >= argc || badopt > 0 || opt_command == NULL)
      {
        usage(0);
        exit(RC_ERROR);
      }

    target_default(opt_rcmd);
    if (opt_maxworkers <= 0)
      {
	fprintf(stderr, "%s: Invalid -M argument!\n", myname);
	exit(RC_ERROR);
      }

    if (opt_spawn == NULL)
	opt_spawn = DEFAULT_SPAWNMODE;

    opt_analyzer = analyzer_init(opt_analyze, opt_outanalysis,opt_erranalysis);
    /* -A requires -o, to avoid dangerous/reckless invocations. */
    if (opt_analyzer != ANALYZE_NONE && opt_odir == NULL)
      {
	fprintf(stderr, "%s: -o option required when using -a/-A!\n", myname);
	exit(RC_ERROR);
      }

    /* -? requires -o, to avoid dangerous/reckless invocations. */
    if ((opt_outmode & (OUT_NULL|OUT_IFERR)) != 0 && opt_odir == NULL)
      {
	fprintf(stderr, "%s: -o option required when using -q!\n", myname);
	exit(RC_ERROR);
      }

    if (opt_odir != NULL)
	opt_outmode |= OUT_COPY;
    else if ((opt_outmode & OUT_ATEND) != 0)
      {
	/* We'll need a temporary directory */
	char *tmp;

	tmp = getenv("TMPDIR");
	snprintf(tdir, sizeof(tdir),
		"%s/%s.%d.%ld", (tmp != NULL) ? tmp : _PATH_TMP, myname,
		(int) getpid(), (long) time(NULL));
	opt_odir = tdir;
      }

    if (opt_odir != NULL && mkdir(opt_odir, 0777) == -1 && errno != EEXIST)
      {
	/* Create odir if it doesn't already exists */
	fprintf(stderr, "%s: mkdir(%s): %s\n",
		myname, opt_odir, strerror(errno));
	exit(RC_ERROR);
      }

    /* Should test outputs be shown? */
    if (opt_vtest > 1)
	opt_test *= -1;

    /* Get list of targets */
    longest = 0;
    while (optind < argc)
      {
	int length;

	if (strcmp(argv[optind], "-") != 0)
          {
	    length = target_add(argv[optind++]);
            if (length > longest)
                longest = length;
          }
	else
	  {
	    char tname[256];

	    optind += 1;
	    while (fgets(tname, 256, stdin) != NULL)
	      {
		tname[strlen(tname)-1] = '\0';
		length = target_add(tname);
                if (length > longest)
                    longest = length;
	      }
	  }
      }
    if (longest == 0)
      {
        fprintf(stderr, "%s: No host given.\n", myname);
        exit(RC_ERROR);
      }
    else if (longest < strlen(myname))
        longest = strlen(myname);
            
    /* Initialize terminal */
    term_init(longest, opt_prefix, opt_status, opt_internal, opt_debug, opt_interactive);

    /* Loop through targets/commands */
    start = time(NULL);
    rc = loop(opt_command, opt_ctimeout, opt_maxworkers, opt_spawn, opt_fail,
	      opt_outmode, opt_odir, opt_analyzer, opt_ping, opt_test);

    /* Summary of results unless asked to be quiet */
    if (opt_quiet == 0)
      {
	nprint("");
	target_results((int) (time(NULL) - start));
      }

    /* odir was temporary, remove it now */
    if (opt_odir != NULL && (opt_outmode & OUT_COPY) == 0
	&& rmdir(opt_odir) == -1)
	fprintf(stderr, "%s: rmdir(%s): %s\n",
		myname, opt_odir, strerror(errno));

    exit(rc);
}
